cylc/flow/job_file.py

```diff
@@ -140,6 +141,18 @@ def _write_header(handle, job_conf):
                  job_conf['execution_time_limit'])]:
             if value:
                 handle.write("\n%s%s" % (prefix, value))
+        # NOTE we cannot do /usr/bin/env bash because we need to use the -l
+        # option and GNU env doesn't support additional arguments (recent
+        # versions permit this with the -S option similar to BSD env but we
+        # cannot make the jump to this until is it more widely adopted
+        handle.write(dedent('''
+
+            if [[ $1 == 'noreinvoke' ]]; then
+                shift
+            else
+                exec "$(command -v bash)" -l "$0" noreinvoke "$@"
+            fi
```

Comment:
@dpmatthews you might want to take a quick look at these lines if you get a chance.

Reply:
(assigning @dpmatthews for 2nd review, based on this comment)

Reply:
I've unassigned as Dave is unlikely to have the time for a full review, its only these particular lines which I think require attention.

Reply:
I think this is OK (we never call the job script with arguments other than "reinvoke" so the shift is redundant but harmless).

It probably doesn't matter but I've been reading about `exec -l` and trying to understand the difference between `exec -l bash` and `exec bash -l`.
https://stackoverflow.com/questions/39972978/option-l-of-exec-shell-command
https://www.gnu.org/software/bash/manual/html_node/Invoking-Bash.html
I'm not sure if there is a difference.

Reply:
On Alpine Linux, `exec -l` doesn't seem to work.

```bash
kinow@ranma:~$ docker run --rm -ti alpine ash
/ # apk add bash
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.11/community/x86_64/APKINDEX.tar.gz
(1/4) Installing ncurses-terminfo-base (6.1_p20200118-r4)
(2/4) Installing ncurses-libs (6.1_p20200118-r4)
(3/4) Installing readline (8.0.1-r0)
(4/4) Installing bash (5.0.11-r1)
Executing bash-5.0.11-r1.post-install
Executing busybox-1.31.1-r9.trigger
OK: 8 MiB in 18 packages
/ # 
/ # exec -l bash
ash: exec: illegal option -l
/ # 
/ # 
/ # exec bash -l
1dfad025d7ce:/# 
```

Though that works on Centos

```
kinow@ranma:~$ docker run --rm -ti centos:centos8 /bin/bash
[root@ef18ef0c049f /]# exec -l bash
[root@ef18ef0c049f /]# exec bash -l
[root@ef18ef0c049f /]# 
```

Ubuntu

```
kinow@ranma:~$ docker run --rm -ti ubuntu:latest /bin/bash
root@41f19db404f1:/# exec -l bash
root@41f19db404f1:/# exec bash -l
root@41f19db404f1:/# 
```

I tried the line with command exec etc on Ubuntu and Alpine, and looks OK to me :+1:

Reply:
(OK, so the form used works everywhere - that we know of :+1: )